[Nguyen Quy Duc] iP

ngquyduc opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen

A project which frees your mind of holding to the tasks you need to do. Duke is
- Text-based: Suitable for people who type fast
- User friendly: easy to remember commands
- **FREE** to use

All you need to do is:
1. Download lastest version of Duke from [here](https://github.com/ngquyduc/ip/releases).
2. Run Duke.
3. Add your tasks and you are good to go.


Try to get yourself familiar with Sheep
- [ ] `todo Do something` to add a todo task 
- [ ] `deadline Deadline /by 2022-08-31` to add add deadline 
- [ ] `event Event /at 2022-08-31` to add an event 
- [ ] `list` to list out all the tasks 
- [ ] `mark 1` to mark the 1st task as done 
- [ ] `unmark 1` to unmark the 1st task 
- [ ] `delete 1` to delete the 1st task 
- [ ] `bye` to close Duke and store all ongoing tasks :wave:

If you want to load a list of tasks, you can load it by replacing ./data/tasks.txt with your file path in main() of file Duke.java

```
public static void main(String[] args) {
    new Duke("./data/tasks.txt").run();
}
```

Hope you have good experience with Duke!!!

Commit:
Add Gradle support

ngquyduc committed:
finish Level 1. Greet, Echo, Exit

ngquyduc committed:
finish Level 2. Add, List

ngquyduc committed:
finish Level 3. Mark as Done

ngquyduc committed:
finish Level 4. ToDos, Events, Deadlines

ngquyduc committed:
finish A-TextUiTesting: Automated Text UI Testing

ngquyduc committed:
finish Level 5. Handle Errors

ngquyduc committed:
finish Level 6. Delete

ngquyduc committed:
save tasks after closing Sheep

ngquyduc committed:
implement dates and time

ngquyduc committed:
Merge branch 'branch-level-7'

ngquyduc committed:
Merge branch 'branch-level-8'

ngquyduc committed:
further implement Dates and Times after merging

ngquyduc committed:
make the code more OOP

ngquyduc committed:
Divide classes into packages

ngquyduc committed:
Merge branch 'add-gradle-support' into Master

ngquyduc committed:
Run Checkstyle Æwith Gradle

ngquyduc committed:
Add JUnit tests

ngquyduc committed:
Package the App as a JAR file

ngquyduc committed:
Add JavaDoc comments to the code

ngquyduc committed:
Fix coding styles

ngquyduc committed:
Implement find for duke

ngquyduc committed:
Merge branch 'branch-A-JavaDoc' into Master

ngquyduc committed:
Merge branch 'branch-A-CodingStandard' into Master

ngquyduc committed:
Merge and resolve conflicts

dexter-sim commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,74 @@
+
+    /**
+     * Unmark a marked task.
+     */
+    public void unmarkDone() {
+        this.done = false;
+    }
+
+    /**
+     * Get the description of a task.
+     * @return
+     */
+    public String getContent() {
+        return this.content;
+    }
+
+    /**
+     * Check if a task description contains given keyword.
+     * @param keyword
+     * @return boolean value
+     */
+    public boolean match(String keyword) {
+        return this.content.contains(keyword);
+    }
+
+    /**
+     * Return 1 if the task is done, 0 otherwise.
+     * @return
+     */
+    public String isDoneString() {
+        return this.done == true ? "1" : "0";
+    }
+
+    /**
+     * String representative of a Task.
+     * @return
+     */
+    @Override
+    public String toString() {
+        return "[" + (this.done ? "X" : " ") + "] " + this.content;
```
The textbook states "Avoid unnecessary use of `this` with fields. Use the `this` keyword only when a field is shadowed by a method or constructor parameter."
Should all `this` be removed in favour of reducing unnecessary noise?

dexter-sim commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,112 @@
+
+    /**
+     * Read the input from the user.
+     * @return
```
Did you forget to write JavaDoc for `@return` or did you left it blank intentionally?

dexter-sim commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+     */
+    @Override
+    public String toString() {
+        String str = "";
+        for (int i = 0; i < getNumberOfTasks(); i++) {
+            String count = (i + 1) + ". ";
+            if (i == 0) {
+                str += count + tasks.get(i).toString();
+            } else {
+                str += "\n" + count + tasks.get(i).toString();
+            }
+        }
+        return str;
```
Would using `StringBuilder` instead of string concatenation improve code efficiency and readability?

dexter-sim commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,98 @@
+     */
+    public ArrayList<Task> load() throws DukeException {
+        File file = new File(filePath);
+        try {
+            Scanner sc = new Scanner(file);
+            ArrayList<Task> tasks = new ArrayList<>();
+            while (sc.hasNext()) {
+                String storedTask = sc.nextLine();
+                String[] splitTask = storedTask.split(" \\| ");
+                Task task;
+                if (splitTask[0].equals("T")) {
+                    task = new Todo(splitTask[2]);
+                } else if (splitTask[0].equals("D")) {
+                    if (DateParser.isDateValid(splitTask[2])) {
+                        task = new Deadline(splitTask[2], DateParser.convertToLocalDate(splitTask[2]));
+                    } else {
+                        System.out.println("OOPS!!! Look like your date format is incorrect");
+                        continue;
+                    }
+                } else {
+                    task = new Event(splitTask[2], splitTask[3]);
+                }
+                if (splitTask[1].equals("1")) {
+                    task.markDone();
+                }
+                tasks.add(task);
+            }
+            return tasks;
+        } catch (FileNotFoundException e) {
```
The textbook states that Avoid Deep Nesting can improve code quality.
Do you think it will be better to reduce to no more than 3 levels of indentation?

dexter-sim commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,36 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (at: " + this.at + ")";
```
Would using `String.format` instead of string concatenation improve code quality?
```java
return String.format("[E]%s (at: %s)", super.toString(), at);
```

dexter-sim reviewed (commented):
Looks good.
Some minor coding standard violations and some code quality that could be improved upon.

ngquyduc committed:
Update README.md

ngquyduc committed:
Fix minor style error

yunruu reviewed (commented):
For JavaDoc, take note that coding standard is to have a spacing between the description and @param :)

yunruu reviewed (commented):
For JavaDoc, take note that coding standard is to have a spacing between the description and @param :)

ngquyduc committed:
Set up JavaFX

ngquyduc committed:
Adding GUI to Duke

ngquyduc committed:
Customize dialog box

ngquyduc committed:
Merge branch 'master' of https://github.com/ngquyduc/ip into Master

ngquyduc committed:
Generate a new JAR file v0.2

ngquyduc committed:
Minor improve GUI for Duke

ngquyduc committed:
Use Assertions

Classes do not have any checks for their invariants or assumptions made.

Unchecked invariants or assumptions may cause the program to crash or
give our incorrect results.

Add assertions to check invariants.

Assertion is preferred as it gets the job done and it can be turned on
and off.

ngquyduc committed:
Merge pull request #1 from ngquyduc/branch-A-Assertions

Use Assertions

ngquyduc committed:
Improve code quality

Some classes have methods that are not in used.

Those methods pile up and make it difficult for other people to read the
code.

Remove unnecessary fields and methods in classes.

Clean the code is better than comment out because the program no longer
relies on those methods and fields.

ngquyduc committed:
Merge pull request #2 from ngquyduc/branch-A-CodeQuality

Improve code quality

ngquyduc committed:
Add tagging for tasks

ngquyduc committed:
Merge branch 'branch-C-Tagging'

ngquyduc committed:
Fix some styling errors

ngquyduc committed:
Reimplement Tag for Task

ngquyduc committed:
Merge branch 'branch-C-Tagging'

ngquyduc committed:
Implement date time for Event and handle more cases in Parser

ngquyduc committed:
Update screenshot

ngquyduc committed:
Update screenshot

ngquyduc committed:
Add a User guide

ngquyduc committed:
Merge branch 'branch_UG'

ngquyduc committed:
Restore the original README.md

ngquyduc committed:
Correct minor spelling error

ngquyduc committed:
Close the main window after typing in 'bye'

ngquyduc committed:
Fix problem with .jar file

ngquyduc committed:
Close the window with button now store the tasks automatically

ngquyduc committed:
Make find feature to work with tag

ngquyduc committed:
Update UG

ngquyduc committed:
Fix minor error in storage for marked tasks
